File: CS/AddRemoveRows/MainViewModel.cs
using DevExpress.Mvvm;
using System.Collections.Generic;
using System.Collections.ObjectModel;

namespace AddRemoveRows {
    public class Product {
        public string ProductName { get; set; }
        public string CompanyName { get; set; }
        public int UnitPrice { get; set; }
        public bool Discontinued { get; set; }

        public Product() { }
        public Product(int i) {
            ProductName = "ProductName" + i;
            CompanyName = "CompanyName" + i;
            UnitPrice = i;
            Discontinued = i % 2 == 0;
        }
    }
    public class MainViewModel : ViewModelBase {
        public ObservableCollection<Product> Products { get; private set; }
        public MainViewModel() {
            Products = new ObservableCollection<Product>(GetSource());
        }
        static IEnumerable<Product> GetSource() {
            for(int i = 0; i < 9; i++) {
                yield return new Product(i);
            }
        }
    }
}

File: VB/AddRemoveRows/MainViewModel.vb
Imports DevExpress.Mvvm
Imports System.Collections.Generic
Imports System.Collections.ObjectModel

Namespace AddRemoveRows

    Public Class Product

        Public Property ProductName As String

        Public Property CompanyName As String

        Public Property UnitPrice As Integer

        Public Property Discontinued As Boolean

        Public Sub New()
        End Sub

        Public Sub New(ByVal i As Integer)
            ProductName = "ProductName" & i
            CompanyName = "CompanyName" & i
            UnitPrice = i
            Discontinued = i Mod 2 = 0
        End Sub
    End Class

    Public Class MainViewModel
        Inherits ViewModelBase

        Private _Products As ObservableCollection(Of AddRemoveRows.Product)

        Public Property Products As ObservableCollection(Of Product)
            Get
                Return _Products
            End Get

            Private Set(ByVal value As ObservableCollection(Of Product))
                _Products = value
            End Set
        End Property

        Public Sub New()
            Products = New ObservableCollection(Of Product)(GetSource())
        End Sub

        Private Shared Iterator Function GetSource() As IEnumerable(Of Product)
            For i As Integer = 0 To 9 - 1
                Yield New Product(i)
            Next
        End Function
    End Class
End Namespace
